### PR TITLE
8338937: Optimize the string concatenation of ClassDesc

### PR DESCRIPTION
**Pull request by wenshao**
The string concatenation of java.base module is implemented based on StringBuilder, which will result in extra object allocation and slow performance. We can solve this problem by using String.concat method and StringConcatHelper to provide concat method.

for example, 

* use "+"
```java
class ClassDesc {
    default String displayName() {
        c.displayName() + "[]".repeat(depth)
    }
}
```

bytecode:
```
106: new           #40                 // class java/lang/StringBuilder
109: dup
110: invokespecial #42                 // Method java/lang/StringBuilder."<init>":()V
113: aload_2
114: invokeinterface #195,  1          // InterfaceMethod displayName:()Ljava/lang/String;
119: invokevirtual #50                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
122: ldc           #198                // String []
124: iload_1
125: invokevirtual #200                // Method java/lang/String.repeat:(I)Ljava/lang/String;
128: invokevirtual #50                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
131: invokevirtual #53                 // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
134: areturn
```

* use String#concat
```java
c.displayName().concat("[]".repeat(depth))
```

bytecode:
```
112: ldc           #198                // String []
114: iload_1
115: invokevirtual #200                // Method java/lang/String.repeat:(I)Ljava/lang/String;
118: invokevirtual #86                 // Method java/lang/String.concat:(Ljava/lang/String;)Ljava/lang/String;
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8338937](https://bugs.openjdk.org/browse/JDK-8338937): Optimize the string concatenation of ClassDesc (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20705/head:pull/20705` \
`$ git checkout pull/20705`

Update a local copy of the PR: \
`$ git checkout pull/20705` \
`$ git pull https://git.openjdk.org/jdk.git pull/20705/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20705`

View PR using the GUI difftool: \
`$ git pr show -t 20705`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20705.diff">https://git.openjdk.org/jdk/pull/20705.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20705#issuecomment-2308850444)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/StringConcatHelper.java [hunk 1]
```diff
@@ -783,4 +783,16 @@ static int checkOverflow(int value) {
         }
         throw new OutOfMemoryError("Overflow: String length out of range");
     }
+
+    static String concat(String prefix, String value, String suffix) {
```
[the remark this answers is not in the document]
Comment (wenshao) [reply src/java.base/share/classes/java/lang/StringConcatHelper.java:787]: In this case, there are too many locations under the java.lang package to change, and a new PR may be more appropriate.

[the remark this answers is not in the document]
Comment (ExE-Boss) [reply src/java.base/share/classes/java/lang/StringConcatHelper.java:787]: Well, `Class​::descriptorString()` is referenced from the `ReferenceClassDescImpl` creation code in `ConstantUtils`:
https://github.com/openjdk/jdk/blob/5671f836039ef1683e3e9ce5b7cf0fa2f1860e2d/src/java.base/share/classes/jdk/internal/constant/ConstantUtils.java#L89-L91

[the remark this answers is not in the document]
Comment (wenshao) [reply src/java.base/share/classes/java/lang/StringConcatHelper.java:787]: I plan to change all of these in another PR https://github.com/wenshao/jdk/commit/b500348cde84e10dc968c7503b8c1dac9131f146

[the remark this answers is not in the document]
Comment (wenshao) [reply src/java.base/share/classes/java/lang/StringConcatHelper.java]: I've accepted your suggestion and don't want to change too much at once.

**Comment by liach** [comment]
For context, some core library developers are currently investigating if we can mark some classes and methods as used in early startup, after this problem was raised in OpenJDK committer's workshop (and encountered in the generics experiment work shown in JVMLS 2024); those methods can fail fast if they use lambdas, and we can make compiler generate regular string concat in other methods.

However, most of your changes here are on code used before string concat is ready, so this is a good improvment.

**Review by liach — APPROVED** [review]
Tier 1-3 tests look good.

**Pull request closed without merging**